crates/net/common/src/bandwidth_monitor.rs

```diff
@@ -0,0 +1,176 @@
+        })
+    }
+
+    /// Returns the total number of bytes that have been downloaded on all the streams.
+    ///
+    /// > **Note**: This method is by design subject to race conditions. The returned value should
+    /// >           only ever be used for statistics purposes.
+    pub fn total_inbound(&self) -> u64 {
+        self.inbound.load(Ordering::Relaxed)
+    }
+
+    /// Returns the total number of bytes that have been uploaded on all the streams.
+    ///
+    /// > **Note**: This method is by design subject to race conditions. The returned value should
+    /// >           only ever be used for statistics purposes.
+    pub fn total_outbound(&self) -> u64 {
+        self.outbound.load(Ordering::Relaxed)
+    }
+}
+
+/// Wraps around a single stream that implements [`AsyncRead`] + [`AsyncWrite`] and monitors the bandwidth through it
+#[pin_project::pin_project]
+pub(crate) struct MonitoredStream<S> {
+    /// The stream this instruments
+    #[pin]
+    inner: S,
+    /// The [`BandwidthMonitor`] struct this uses to monitor bandwidth
+    monitor: Arc<BandwidthMonitor>,
+}
+
+impl<S> MonitoredStream<S> {
+    fn new(inner: S) -> Self {
+        Self {
+            inner,
+            monitor: BandwidthMonitor::new()
+        }
+    }
+}
+
+impl<Stream: AsyncRead> AsyncRead for MonitoredStream<Stream> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let this = self.project();
+        let num_bytes = {
+            let init_num_bytes = buf.filled().len();
+            ready!(this.inner.poll_read(cx, buf))?;
+            buf.filled().len() - init_num_bytes
+        };
+        this.monitor.inbound.fetch_add(
+            u64::try_from(num_bytes).unwrap_or(u64::max_value()),
+            Ordering::Relaxed,
+        );
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<Stream: AsyncWrite> AsyncWrite for MonitoredStream<Stream> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
+        let num_bytes = ready!(this.inner.poll_write(cx, buf))?;
+        this.monitor.outbound.fetch_add(
+            u64::try_from(num_bytes).unwrap_or(u64::max_value()),
+            Ordering::Relaxed,
+        );
+        Poll::Ready(Ok(num_bytes))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let this = self.project();
+        this.inner.poll_flush(cx)
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let this = self.project();
+        this.inner.poll_shutdown(cx)
+    }
+}
+
+// TODO: Something similar to rust-libp2p's `StreamMuxer` to monitor bandwidth across _all_ streams?
+
+// TODO: Tests
+// - Test reading + writing to/from a basic stream
+// - Test reeading + writing to/from a TCP stream
```

Comment:
that'd be nice to have basically server client, send some data and compare read==write

Reply:
my test with a basic `DuplexStream` (pipe) essentially does this, do we still want a separate test w/ a TCP stream?

Reply:
Added a TcpStream test as well! lmk if that's what you were looking for